### PR TITLE
[SPARK-33655][SQL] Improve performance of processing FETCH_PRIOR

### PR DESCRIPTION
**Pull request by Dooyoung-Hwang**
### What changes were proposed in this pull request?
Currently, when a client requests FETCH_PRIOR to Thriftserver, Thriftserver reiterates from the start position. Because Thriftserver caches a query result with an array when THRIFTSERVER_INCREMENTAL_COLLECT feature is off, FETCH_PRIOR can be implemented without reiterating the result. A trait FeatureIterator is added in order to separate the implementation for iterator and an array. Also, FeatureIterator supports moves cursor with absolute position, which will be useful for the implementation of FETCH_RELATIVE, FETCH_ABSOLUTE.

### Why are the changes needed?
For better performance of Thriftserver.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
FetchIteratorSuite

### CODE REVIEW

**Comment by Dooyoung-Hwang** [comment]
@juliuszsompolski
I created PR.

cc : @HyukjinKwon , @dongjoon-hyun , @cloud-fan

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @wangyum as well

**Comment by Dooyoung-Hwang** [comment]
Jenkins, retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Review by juliuszsompolski — APPROVED** [review]
Looks good to me.
Thanks @Dooyoung-Hwang

**Review by HyukjinKwon — COMMENTED** [review]
Looks fine to me too. @wangyum can you do a final check when you find some time?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 1]
```diff
@@ -148,43 +142,10 @@ private[hive] class SparkExecuteStatementOperation(
...
-      } else { // FETCH_NEXT
-        previousFetchEndOffset
-      }
-    }
-
-    resultRowSet.setStartOffset(resultOffset)
-    previousFetchStartOffset = resultOffset
+    if (order.equals(FetchOrientation.FETCH_FIRST)) iter.fetchAbsolute(0)
+    else if (order.equals(FetchOrientation.FETCH_PRIOR)) iter.fetchPrior(maxRowsL)
+    else iter.fetchNext()
```
Comment (wangyum) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:147]: Add curly braces? https://github.com/databricks/scala-style-guide/blob/master/README.md#curly
cc @maropu

**Review by HyukjinKwon — APPROVED** [review]

**Comment by wangyum** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by Dooyoung-Hwang** [comment]
Thanks!!

**Pull request closed without merging**